pkg/codegen/dotnet/gen_program_test.go

```diff
@@ -1,11 +1,97 @@
 package dotnet
 
 import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
 	"testing"
 
+	"github.com/stretchr/testify/assert"
+
 	"github.com/pulumi/pulumi/pkg/v3/codegen/internal/test"
+	"github.com/pulumi/pulumi/pkg/v3/testing/integration"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/executable"
 )
 
 func TestGenerateProgram(t *testing.T) {
-	test.TestProgramCodegen(t, "dotnet", GenerateProgram)
+	test.TestProgramCodegen(t,
+		test.ProgramLangConfig{
+			Language:   "dotnet",
+			Extension:  "cs",
+			OutputFile: "MyStack.cs",
+			Check:      checkDotnet,
+			GenProgram: GenerateProgram,
+		},
+	)
+}
+
+func checkDotnet(t *testing.T, path string) {
+	var err error
+	dir := filepath.Dir(path)
+
+	ex, err := executable.FindExecutable("dotnet")
+	assert.NoError(t, err, "Failed to find dotnet executable")
+
+	projectFile := filepath.Join(dir, filepath.Base(dir)+".csproj")
+	if _, err := ioutil.ReadFile(projectFile); os.IsNotExist(err) {
+		defer func() {
+			err = os.Remove(projectFile)
+			assert.NoError(t, err, "Failed to delete project file")
+			err = os.Remove(filepath.Join(dir, "Program.cs"))
+			assert.NoError(t, err, "Failed to delete C# project main")
+		}()
+		err = integration.RunCommand(t, "create dotnet project",
+			[]string{ex, "new", "console"}, dir, &integration.ProgramTestOptions{})
+		assert.NoError(t, err, "Failed to create C# project")
+	}
+
+	// Add dependencies (based on directory name)
+	if pkg, pkgVersion := packagesFromTestName(filepath.Base(dir)); pkg != "" {
+		err = integration.RunCommand(t, "create dotnet project",
+			[]string{ex, "add", "package", pkg, "--version", pkgVersion},
+			dir, &integration.ProgramTestOptions{})
+		assert.NoError(t, err, "Failed to add dependency %q %q", pkg, pkgVersion)
+	} else {
+		err = integration.RunCommand(t, "add sdk ref",
+			[]string{ex, "add", "reference", "../../../../../../sdk/dotnet/Pulumi"},
```

Comment:
This is nice! I'd want that in SDK codegen tests also, to link against in-repo Pulumi SDK.

Reply:
We can do that, but it's slightly more complicated. In the SDK codegen comp tests we rely on the SDK generator to build the package (including it's metadata). My desire was to avoid modifying the generated package metadata. There it's a trade off, changing what part of the system is put under test. Here we need to manually build the metadata (so we may as well do it right).